src/tests/Common/Directory.Build.targets

```diff
@@ -34,7 +35,47 @@
+        instead of `-debug -p:RuntimeConfiguration=checked`. That forces us to either pass the
+        `HostConfiguration=debug` explicitly or fix this disparity; both of which will break the dev workflow.
+        
+        As a workaround, we will first check if the directory pointed by `HostConfiguration` exists, then check
+        `LibrariesConfiguration`.
+    -->
+
+    <PropertyGroup Condition="'$(Crossgen2Supported)' == 'true'">
+      <_targetOS>$(TargetOS)</_targetOS>
+      <_targetOS Condition="'$(_targetOS)' == 'windows'">win</_targetOS>
+      <_apphostPath Condition="Exists('$(ArtifactsBinDir)$(_targetOS)-$(TargetArchitecture).$(HostConfiguration)\corehost\apphost$(ExeSuffix)')">$(ArtifactsBinDir)$(_targetOS)-$(TargetArchitecture).$(HostConfiguration)\corehost\apphost$(ExeSuffix)</_apphostPath>
+      <_apphostPath Condition="'$(_apphostPath)' == '' and Exists('$(ArtifactsBinDir)$(_targetOS)-$(TargetArchitecture).$(LibrariesConfiguration)\corehost\apphost$(ExeSuffix)')">$(ArtifactsBinDir)$(_targetOS)-$(TargetArchitecture).$(LibrariesConfiguration)\corehost\apphost$(ExeSuffix)</_apphostPath>
+      <_toolsConfiguration Condition="Exists('$(ArtifactsBinDir)ILLink.Tasks\$(ToolsConfiguration)\$(NetCoreAppToolCurrent)\ILLink.Tasks.dll')">$(ToolsConfiguration)</_toolsConfiguration>
+      <_toolsConfiguration Condition="Exists('$(ArtifactsBinDir)ILLink.Tasks\$(LibrariesConfiguration)\$(NetCoreAppToolCurrent)\ILLink.Tasks.dll')">$(LibrariesConfiguration)</_toolsConfiguration>
+    </PropertyGroup>
+
+    <MakeDir Condition="'$(Crossgen2Supported)' == 'true'" Directories="$(ArtifactsObjDir)coreclr\crossgen2_publish\$(TargetOS).$(TargetArchitecture).$(RuntimeConfiguration)" />
+
+    <Copy
+      SourceFiles="$(_apphostPath)"
+      DestinationFiles="$(ArtifactsObjDir)coreclr\crossgen2_publish\$(TargetOS).$(TargetArchitecture).$(RuntimeConfiguration)\apphost$(ExeSuffix)"
+      Condition="'$(Crossgen2Supported)' == 'true'" />
+
+    <!-- Publish crossgen2 on supported platforms. -->
+
+    <MSBuild Condition="'$(Crossgen2Supported)' == 'true'"
+      Targets="Restore"
+      BuildInParallel="true"
+      Properties="NativeAotSupported=$(NativeAotSupported);UseNativeAotForComponents=$(UseNativeAotForComponents);ToolsConfiguration=$(_toolsConfiguration);MSBuildRestoreSessionId=$([System.Guid]::NewGuid());"
+      Projects="$(InstallerProjectRoot)pkg\sfx\Microsoft.NETCore.App\Microsoft.NETCore.App.Crossgen2.sfxproj" />
+
+    <MSBuild Condition="'$(Crossgen2Supported)' == 'true'"
+      Targets="PublishToDisk"
+      BuildInParallel="true"
+      Properties="NativeAotSupported=$(NativeAotSupported);UseNativeAotForComponents=$(UseNativeAotForComponents);ToolsConfiguration=$(_toolsConfiguration);OutputPath=$(CORE_ROOT)\crossgen2;LibrariesConfiguration=$(LibrariesConfiguration)"
+      Projects="$(InstallerProjectRoot)pkg\sfx\Microsoft.NETCore.App\Microsoft.NETCore.App.Crossgen2.sfxproj" />
```

Comment:
Is this what invokes ilc to AOT compile crossgen2? It seems unexpected to me to do that from `src/tests/build.sh`, especially when `generatelayoutonly` is passed. Rather it seems like this AOT'ing ought to be done as part of invoking the `build.sh` in the repo root.

Reply:
crossgen2.exe is required by core_root layout in some contexts. We were relying on crossgen2.dll from the artifacts in those cases, now we will be using the same form-factor in tests which we are shipping in the crossgen2 nuget package (https://github.com/dotnet/runtime/issues/80110).

Reply:
The documentation for `generatelayoutonly` explicitly says it does not build managed or native test components, so it seems odd to me to start doing that.

Reply:
It produces CORE_ROOT with crossgen2.dll? This is part of that.

Reply:
`generatelayoutonly` is meant to copy prebuilt bits to create `Core_Root`. I do not think it should be running ilc. If we have a build step needed for tests, then that should happen as part of the build.

Reply:
I have changed it already. The pipelines which require crossgen.exe have it running, others are skipping this step.

>If we have a build step needed for tests, then that should happen as part of the build.

Unfortunately, this is not possible. The publishing happen as part of the "packs" subset today. That subset is not built in many validation pipelines.

Reply:
Hmm, ok. 

I was going to mention that this has the same flavor as ilc-published that is copied to Core_Root. ilc-published is also the NAOT compiled version of ilc. Can crossgen2 not be handled similarly to it?

Reply:
ilc uses nuget package to publish, crossgen2 uses live ilc build for publishing. In theory, we can bring them to the same live-build plan (i.e. use live ilc.dll to publish ilc, then use that ilc to publish again to close the loop), but it requires some work. I'm figuring out similar issues with ILCompiler.csproj for FreeBSD (where ilc is PublishSingleFile) https://github.com/dotnet/runtime/pull/105004.

Reply:
Ok, thanks. Then let's keep it like this.